src/main/java/de/tum/in/www1/artemis/service/hestia/ExerciseHintService.java

```diff
@@ -0,0 +1,267 @@
+            // current task is successful
+            if (subsequentNumberOfUnsuccessfulSubmissionsForCurrentTask == 0) {
+                continue;
+            }
+
+            var hintsInTask = exerciseHints.stream()
+                    .filter(hint -> Objects.nonNull(hint.getProgrammingExerciseTask()) && Objects.equals(hint.getProgrammingExerciseTask().getId(), task.getId()))
+                    .collect(Collectors.toSet());
+            // no hints exist for the current task
+            if (hintsInTask.isEmpty()) {
+                continue;
+            }
+
+            Optional<Integer> subsequentNumberSuccessfulSubmissionsForPreviousTask;
+            if (i == 0) {
+                subsequentNumberSuccessfulSubmissionsForPreviousTask = Optional.empty();
+            }
+            else {
+                subsequentNumberSuccessfulSubmissionsForPreviousTask = Optional.of(subsequentNumberOfSuccessfulSubmissionsByTask.get(tasks.get(i - 1)));
+            }
+
+            // skip current task if the previous task was not successful
+            if (Objects.equals(subsequentNumberSuccessfulSubmissionsForPreviousTask.orElse(-1), 0)) {
+                continue;
+            }
+
+            // add the available hints for the current task
+            var availableHintsForCurrentTask = getAvailableExerciseHintsForTask(subsequentNumberSuccessfulSubmissionsForPreviousTask,
+                    subsequentNumberOfUnsuccessfulSubmissionsForCurrentTask, hintsInTask);
+            if (availableHintsForCurrentTask.size() > 0) {
+                return availableHintsForCurrentTask;
+            }
+        }
+        return new HashSet<>();
+    }
+
+    private boolean isTaskSuccessfulInSubmission(ProgrammingExerciseTask task, Submission submission) {
+        var result = submission.getLatestResult();
+        if (result == null || result.getFeedbacks().isEmpty()) {
+            return false;
+        }
+        var testCasesInTask = task.getTestCases();
+        var feedbacks = result.getFeedbacks();
+        return feedbacks.stream().filter(feedback -> testCasesInTask.stream().anyMatch(testCase -> Objects.equals(testCase.getTestName(), feedback.getText())))
+                .allMatch(Feedback::isPositive);
+    }
+
+    private List<Submission> getSubmissionsForStudent(ProgrammingExercise exercise, User student) {
+        var studentParticipation = studentParticipationRepository.findByExerciseIdAndStudentIdWithEagerSubmissionsResultsFeedbacks(exercise.getId(), student.getId())
+                .orElseThrow(() -> new InternalServerErrorException("No participation"));
+        return studentParticipation.getSubmissions().stream().sorted(Comparator.comparing(Submission::getSubmissionDate, Comparator.reverseOrder())).toList();
+    }
+
+    private int subsequentNumberOfSuccessfulSubmissionsForTask(List<Submission> submissions, ProgrammingExerciseTask task, boolean successful) {
```

Comment:
The method name containing "Successful" and the parameter successful are a bit confusing